### DISCUSSION TITLE
Best way to deduplicate identical rows

### DISCUSSION THREAD

**Discussion by roengram** [ClickHouse/ClickHouse#64554] [Q&A] [answered]
Hi, the title is pretty much everything I'm dying to find out.

System is simple: an API server receives request and ingest the content into CH. What makes this complicated is that clients can sometimes send duplicate requests in situations like retransmission, leading to the duplicate rows in CH. Our business logic does not allow duplicate rows.

**Environment**
 - Use ClickHouse server 24.4.1.2088
 - Client can send duplicate rows with an interval of several days or even weeks
 - The number of deduplicated rows is approximately 1B/day
 - Table engine is ReplicatedMergeTree, having 3 replica (hosts) with 1 shard
 - Table has about 20 columns, mostly String but also several Map, DateTime, and Int64. Many of them are high cardinality. About 15 columns are dimensions, meaning that their combination must be checked for duplication
 - Duplicates are expected to be very rare, less than 0.001%. So it's best to have a dedup strategy with low overheads for most of normal cases.

**Option 1: use built-in block deduplication**
After some investigation, I came across with this function: [insert_deduplicate](https://clickhouse.com/docs/en/operations/settings/settings#insert-deduplicate). Actually, the functionality is exactly what I want - simply ignore duplicate rows. But unfortunately the time window and number of recent blocks to check duplicates are too low for our workload.

**Option 2: perform SELECT DISTINCT before aggregation**
This is actually the current interim solution. Currentlly we have only test data so SELECT DISTINCT is not a problem. But I'm worried if this approach will not scale well when the service rolls out in production. I thought of creating a specific dedup key column, which stores a hash of all other columns, to lower the overhead of performing DISTINCT. I'm not sure whether this column will actually help reducing cost and increasing performance though.

**Option 3: use AggregatingMergeTree**
I'm just beginning to poke around this approach in hope to incrementally deduplicate. But I could not find appropriate aggregate function from [the supported aggregate functions list](https://clickhouse.com/docs/en/sql-reference/aggregate-functions/reference). Is there anything I missed?

**Option 4: use GROUP BY**
Search results include using GROUP BY. This is doable but I do not know if it is more or performant/efficient than SELECT DISTINCT.

So, I'm asking help from the community. Any suggestion will be welcome.

**Comment by yiyang-shao** [comment]
Hi, the options and your understanding is correct. In addition, check out [ReplacingMergeTree](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replacingmergetree)

**Comment by roengram** [comment]
@yiyang-shao Thank you very much for your quick reply.

I've investigated ReplacingMergeTree, and have some questions:


**1. Is there any way to guarantee to have a deduplicated table?**

I understand that dedup is eventual and not predictable; dedup happens when merging, which is unpredictably scheduled as background task unless user queries `OPTIMIZE` or `FINAL`. Executing `FINAL` is not a problem for me, but I found references saying that even that  does not guarantee deduplication:

  > Merges are eventual and may never happen. It depends on the number of inserts that happened after, the number of parts in the partition, size of parts. If the total size of input parts are greater than the maximum part size then they will never be merged. ([ref](https://kb.altinity.com/engines/mergetree-table-engine-family/replacingmergetree/altinity-kb-replacingmergetree-does-not-collapse-duplicates/))

Probably the above is not relevant for my case. In my data, all duplicate rows are in the same partition; i.e. I can use `do_not_merge_across_partitions_select_final`. But I'd also like to know the cases when `FINAL` fails to deduplicate.


**2. Is there a way to check if merge has been executed?**

Again, I understand the merge itself is asynchronous and done as background task. But can I find out when the merge actually happened? Probably only from logging?
Or, should I actually compare `SELECT COUNT(DISTINCT(*))` and `SELECT COUNT(*)` to be confident?


**3. If I do `FINAL` with `WHERE`, only matching rows are subject to merge?**

I'd like to get deduped query rows except too new ones; e.g. only rows older than 1 week but younger than 5 weeks. In this case, does `FINAL` cause merging only the matching rows?

**Reply by yiyang-shao** [reply, marked as answer]
Hi @roengram 

> **1. Is there any way to guarantee to have a deduplicated table?**

The doc means the `OPTIMIZE` process is not guaranteed, thus causing failed deduplication. However, if merge does not execute properly, it may cause other issues. So it is better to find out why the merge is broken.
To guarantee a deduplicated query result, you may 
1. use `SELECT xxx from <table> FINAL`, but it will have a performance impact on the query
2. run `OPTIMIZE FINAL` periodically and verify the data consistency. 

> **2. Is there a way to check if merge has been executed?**

You are right. The server logs contain the most information. 
However, for current running merges, you can check the `sytem.merges` table([ref](https://clickhouse.com/docs/en/operations/system-tables/merges))
For past merges, you can check the `system.part_log` table([ref](https://clickhouse.com/docs/en/operations/system-tables/part_log)) to get more historical information. E.g.
```sql
SELECT
    event_time,
    part_name,
    merged_from,
    rows,
    bytes
FROM system.part_log
WHERE table = 'my_table' AND event_type = 'Merge'
ORDER BY event_time DESC
LIMIT 10;
```
Besides, comparing `SELECT COUNT(DISTINCT(*))` and `SELECT COUNT(*)`, you can also use the following to check.
```sql
SELECT id, COUNT(*)
FROM my_table
FINAL
GROUP BY id
HAVING COUNT(*) > 1;
```

> **3. If I do FINAL with WHERE, only matching rows are subject to merge?**

It is possible if you partition your table by date or week. 
In `SELECT` query, use `WHERE` to filter by partition. Then only specific partitions will be processed.
You can also optimize a specific partition, using `OPTIMIZE TABLE my_table PARTITION '20240530' FINAL;`

Hope the answer above helps.

**Comment by shubhambhardwaj98** [comment]
Hi 

We also encountered a similar problem at a smaller scale where we are inserting 100M rows daily with info about UserSessions on the app. We initially implemented a `ReplacingMergeTree` Table and were querying our business logic on top of it. 

The issue with `ReplacingMergeTree` is the dedupe operation is triggered at some point in time decided by Clickhouse automatically and not necessarily run before you query.


We observed increases in query times by 100x when using `FINAL` and is typically considered bad as it triggers an extra merge operation during query execution.

So the next simple solution was implementing getting latest states in the query logic itself.

This would look something like this 
```plpgsql
with cte as (
SELECT col1,col2,argMax(col3,updated_at) as latest_col_3
FROM table_name
group by 1,2 )

-- Your business logic on top of the CTE
``` 

This again is a huge overhead what we ended up eventually doing was to make an `AggregatedMergeTree` table. The aggregated column was the one that was used the most in our business logic.

Implementation - We leverage Incremental Materialized Views  [CH Docs](https://clickhouse.com/docs/materialized-view/incremental-materialized-view) to populate a table on top of which business logic is queried upon.

Create an Incremental materialized view - 
```plpgsql
CREATE MATERIALIZED VIEW mv_name
            TO target_table
SELECT col1 ,
       col2,
       uniqState(cityHash64(col3))  as col3
FROM your_original_table
GROUP BY 1,2
``` 

Unlike traditional Materialized views the goal is not to refresh a query on certain schedules or triggers. Incremental MVs are just a trigger that runs a query on blocks of data as they are inserted into your source table the results of which are eventually sent to a target table. 
This is also the step where you want to do your filtering , aggregating and changing base columns in desired formats. A great thing is these Materialized Views are updated in real time as data flows into your  target_table.

We were able to bring down the number of rows to roughly 1/5th of the original size. This was the fastest solution out of all which covered 95% of our usecases.
